Sources/PackageCollections/PackageIndex.swift

```diff
@@ -0,0 +1,186 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import Foundation
+import PackageModel
+import TSCBasic
+
+struct PackageIndex: PackageIndexProtocol {
+    private let configurationStorage: PackageIndexConfigurationStorage
+    private let httpClient: HTTPClient
+    private let callbackQueue: DispatchQueue
+    private let observabilityScope: ObservabilityScope
+    
+    // Feature flag: env var SWIFTPM_ENABLE_PACKAGE_INDEX
+    let enabled: Bool
+
+    // TODO: cache metadata results
+
+    init(
+        fileSystem: FileSystem,
+        customHTTPClient: HTTPClient? = nil,
+        callbackQueue: DispatchQueue,
+        observabilityScope: ObservabilityScope
+    ) {
+        self.configurationStorage = PackageIndexConfigurationStorage(fileSystem: fileSystem)
+        self.httpClient = customHTTPClient ?? HTTPClient()
+        self.callbackQueue = callbackQueue
+        self.observabilityScope = observabilityScope
+        self.enabled = ProcessInfo.processInfo.environment["SWIFTPM_ENABLE_PACKAGE_INDEX"] == "1"
+    }
+
+    func set(
+        url: Foundation.URL,
+        callback: @escaping (Result<Void, Error>) -> Void
+    ) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.update(index: .init(url: url)) { result in
+            callback(result.tryMap { () })
+        }
+    }
+
+    func unset(callback: @escaping (Result<Void, Error>) -> Void) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.update(index: .init(url: nil)) { result in
+            callback(result.tryMap { () })
+        }
+    }
+
+    func get(callback: @escaping (Result<Foundation.URL?, Error>) -> Void) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.get { result in
+            callback(result.tryMap { $0.url })
+        }
+    }
+
+    func getPackageMetadata(
+        identity: PackageIdentity,
+        location: String?,
+        callback: @escaping (Result<PackageCollectionsModel.PackageMetadata, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { _, _ in
+            // TODO: call package index's get metadata API
+            fatalError("Not implemented: \(#function)")
```

Comment:
We should probably have some "fake" implementation at least so we can write tests?

Reply:
yea that is a goo idea, maybe even a real http client with fake routes and payloads

Reply:
https://github.com/apple/swift-package-manager/pull/4010/commits/cd27dc5fc04e4392ca27e8f487634a794cf98c07